File: public/python/constraints/no_cycles.py
from typing import Optional
from pipes_utils import Assignment, PipeType, PartialAssignment
from csp import Variable
from math import sqrt
from pipes_utils import check_connections, find_adj

# Constraint to ensure that there are no cycles in the pipes
# implementation details can be found in the Constraints/no_cycles.py section in the report


def assignment_has_cycle(
    curr: int,
    assignment: Assignment,
    visited: set[int],
    prev: Optional[int] = None,
) -> bool:
    if curr in visited:
        return True
    visited.add(curr)

    adj_indexes = find_adj(curr, int(sqrt(len(assignment))))

    center_pipe = assignment[curr]
    top_pipe, right_pipe, bottom_pipe, left_pipe = [
        assignment[i] if i != -1 else None for i in adj_indexes
    ]

    pipe_tuple = (top_pipe, right_pipe, bottom_pipe, left_pipe)

    adj_connections = check_connections(center_pipe, pipe_tuple)

    for adj_i, adj_is_connected in zip(adj_indexes, adj_connections):
        if adj_is_connected and adj_i != prev:
            if assignment_has_cycle(adj_i, assignment, visited, curr):
                return True
    return False


def validator(assignment: Assignment) -> bool:
    return not assignment_has_cycle(0, assignment, set())


def get_duplicated_touched(
    curr: int,
    assignment: PartialAssignment,
    visited: set[int],
    touched: dict[int, int],
    prev: Optional[int] = None,
) -> Optional[tuple[int, int, int]]:
    """
    Creates a tree from assignment, checks if any squares are touched twice.
    A square is "touched" if a pipe's opening is pointing towards it.

    :param curr: the current square index
    :param assignment: the assignment
    :param touched: the set of touched squares indices
    :param prev: the previous square index

    :return: index of the square touched twice, or None if no square is touched twice
    """
    visited.add(curr)
    center_pipe = assignment[curr]
    if center_pipe is None:
        raise Exception("Traversed to an unassigned pipe")

    adj_indexes = find_adj(curr, int(sqrt(len(assignment))))

    for i, adj_i in enumerate(adj_indexes):
        if center_pipe[i]:
            if adj_i == -1:
                raise Exception(
                    f"Pipe pointing to edge of grid in the direction of {i}"
                )

            if adj_i != prev and adj_i in touched:
                return (adj_i, curr, touched[adj_i])

            touched[adj_i] = curr

    top_pipe, right_pipe, bottom_pipe, left_pipe = [
        assignment[i] if i != -1 else None for i in adj_indexes
    ]
    pipe_tuple = (top_pipe, right_pipe, bottom_pipe, left_pipe)

    adj_connections = check_connections(center_pipe, pipe_tuple)

    for adj_i, adj_is_connected in zip(adj_indexes, adj_connections):
        if adj_is_connected and adj_i != prev:
            duplicate_touch = get_duplicated_touched(
                curr=adj_i,
                assignment=assignment,
                touched=touched,
                visited=visited,
                prev=curr,
            )
            if duplicate_touch:
                return duplicate_touch
    return None


def pruner(variables: list[Variable]) -> dict[Variable, list[PipeType]]:
    assignment: PartialAssignment = [var.get_assignment() for var in variables]
    n = int(sqrt(len(assignment)))

    visited: set[int] = set()
    for assignment_index, assignment_value in enumerate(assignment):
        if assignment_value != None and not assignment_index in visited:
            duplicate_touch = get_duplicated_touched(
                curr=assignment_index,
                assignment=assignment,
                visited=visited,
                touched={},
            )

            if duplicate_touch:
                variable_to_prune = next(
                    (var for var in variables if var.location == duplicate_touch[0]),
                )

                top, right, bottom, left = find_adj(duplicate_touch[0], n)
                directions = [top, right, bottom, left]
                touches = [duplicate_touch[1], duplicate_touch[2]]
                touched_directions: list[int] = []

                for i, direction_value in enumerate(directions):
                    for touch_value in touches:
                        if direction_value == touch_value:
                            touched_directions.append(i)

                pruned_values: list[PipeType] = []
                for active_domain in variable_to_prune.active_domain:
                    if (
                        active_domain[touched_directions[0]]
                        and active_domain[touched_directions[1]]
                    ):
                        pruned_values.append(active_domain)

                pruned_dict: dict[Variable, list[PipeType]] = {
                    variable_to_prune: pruned_values
                }

                variable_to_prune.prune(pruned_values)

                return pruned_dict
    return {}

File: utils/csp/constraints/no_cycles.ts
import { Openings, Assignment, findAdj, checkConnections } from "../utils";
import { Variable } from "../csp";

function assignmentHasCycle(
  curr: number,
  assignment: Assignment,
  visited: Set<number>,
  prev: number | null = null
): boolean {
  if (visited.has(curr)) {
    return true;
  }

  visited.add(curr);
  const adjIndexes = findAdj(curr, Math.sqrt(assignment.length));

  const centerPipe = assignment[curr];
  const pipes = adjIndexes.map((i) => (i !== -1 ? assignment[i] : null));

  const adjConnections = checkConnections(centerPipe, pipes);

  for (let i = 0; i < 4; i++) {
    if (adjConnections[i] && adjIndexes[i] !== prev) {
      if (assignmentHasCycle(adjIndexes[i], assignment, visited, curr)) {
        return true;
      }
    }
  }

  return false;
}

export function validator(assignment: Assignment): boolean {
  return !assignmentHasCycle(0, assignment, new Set<number>());
}

function getDuplicatedTouched(
  curr: number,
  assignment: Array<Openings | null>,
  visited: Set<number>,
  touched: Map<number, number>,
  prev: number | null = null
): [number, number, number] | null {
  visited.add(curr);

  const centerPipe = assignment[curr];
  if (centerPipe === null) {
    throw new Error("Traversed to an unassigned pipe");
  }

  const adjIndexes = findAdj(curr, Math.sqrt(assignment.length));

  for (let i = 0; i < 4; i++) {
    if (centerPipe[i]) {
      if (adjIndexes[i] === -1) {
        throw new Error(
          `Pipe pointing to edge of grid in the direction of ${i}`
        );
      }

      if (adjIndexes[i] !== prev && touched.has(adjIndexes[i])) {
        return [adjIndexes[i], curr, touched.get(adjIndexes[i])!];
      }

      touched.set(adjIndexes[i], curr);
    }
  }

  const pipes = adjIndexes.map((i) => (i !== -1 ? assignment[i] : null));
  const adjConnections = checkConnections(centerPipe, pipes);

  for (let i = 0; i < 4; i++) {
    if (adjConnections[i] && adjIndexes[i] !== prev) {
      const duplicateTouch = getDuplicatedTouched(
        adjIndexes[i],
        assignment,
        visited,
        touched,
        curr
      );

      if (duplicateTouch) {
        return duplicateTouch;
      }
    }
  }

  return null;
}

export function pruner(variables: Variable[]): Map<Variable, Openings[]> {
  const assignment: Array<Openings | null> = variables.map((v) =>
    v.getAssignment()
  );
  const n = Math.sqrt(assignment.length);

  const visited: Set<number> = new Set();
  for (
    let assignmentIndex = 0;
    assignmentIndex < assignment.length;
    assignmentIndex++
  ) {
    const assignmentValue = assignment[assignmentIndex];

    if (assignmentValue !== null && !visited.has(assignmentIndex)) {
      const duplicateTouch = getDuplicatedTouched(
        assignmentIndex,
        assignment,
        visited,
        new Map<number, number>()
      );

      if (duplicateTouch) {
        const variableToPrune = variables.find(
          (v) => v.location === duplicateTouch[0]
        );

        if (!variableToPrune) continue;

        const [top, right, bottom, left] = findAdj(duplicateTouch[0], n);
        const directions = [top, right, bottom, left];
        const touches = [duplicateTouch[1], duplicateTouch[2]];
        const touchedDirections: number[] = [];

        for (let i = 0; i < directions.length; i++) {
          for (const touchValue of touches) {
            if (directions[i] === touchValue) {
              touchedDirections.push(i);
            }
          }
        }

        const prunedValues: Openings[] = [];
        for (const activeDomain of variableToPrune.getActiveDomain()) {
          if (
            activeDomain[touchedDirections[0]] &&
            activeDomain[touchedDirections[1]]
          ) {
            prunedValues.push(activeDomain);
          }
        }

        const prunedDict = new Map<Variable, Openings[]>();
        prunedDict.set(variableToPrune, prunedValues);

        variableToPrune.prune(prunedValues);

        return prunedDict;
      }
    }
  }

  return new Map<Variable, Openings[]>();
}
